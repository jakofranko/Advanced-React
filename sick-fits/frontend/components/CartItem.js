import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import RemoveCartItem from './RemoveCartItem';
import formatMoney from '../lib/formatMoney';

const CartItemStyles = styled.li`
    padding: 1rem 0;
    border-bottom:  1px solid ${props => props.theme.lightgrey};
    display: grid;
    align-items: center;
    grid-template-columns: auto 1fr auto;
    img {
        margin-right: 10px;
    }
    h3,
    p {
        margin: 0;
    }
`;

class CartItem extends Component {
    static propTypes = {
        cartItem: PropTypes.object.isRequired,
    }
    render() {
        const { cartItem } = this.props;
        const { item } = cartItem;
        return (
            <CartItemStyles>
                {!item && <p>This item has been removed.</p>}
                <img width="100" src={item.image} alt={item.title} />
                <div className="cart-item-details">
                    <h3>{item.title}</h3>
                    <p>
                        {formatMoney(item.price * cartItem.quantity)}
                        {' - '}
                        <em>
                            {cartItem.quantity} &times; {formatMoney(item.price)}
                        </em>
                    </p>
                </div>
                <RemoveCartItem id={cartItem.id} />
            </CartItemStyles>
        );
    }

}

export default CartItem;
