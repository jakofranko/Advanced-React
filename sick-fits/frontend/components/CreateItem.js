// This component is responsible for sending data to
// our graph QL Yoga server. It does this via the Mutation
// component provided by Apollo (react-apollo).
// The `gql` function allows us to create a mutation that
// will be used by the Mutation component. The Apollo engineers
// also created graphql-tag. In the query that we wrote, we are
// referencing the createItem query that exists in our GraphQL
// backend. Referencing a query like this requires knowledge of
// the GraphQL schema that has been authored in our backend.
// The Mutation component takes our GraphQL mutation as a prop,
// and then exposes it to its only expected child: a render
// component. So, we wrap our entire form in this anonymous
// render component, and invoke the mutation in our onSubmit
// event handler. This function executes the mutation, awaits
// the response, and then re-routes the application. Strangely,
// the mutation function that is exposed by the Mutation component
// doesn't take any arguments; it seems to take the entire state of the component implicitely as its args.

import React, { Component } from 'react';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import Router from 'next/router';
import Form from './styles/Form';
import ErrorMessage from './ErrorMessage';
import formatMoney from '../lib/formatMoney';

export const CREATE_ITEM_MUTATION = gql`
    mutation CREATE_ITEM_MUTATION(
        $title: String!
        $description: String!
        $price: Int!
        $image: String
        $largeImage: String
    ) {
        createItem(
            title: $title
            description: $description
            price: $price
            image: $image
            largeImage: $largeImage
        ) {
            id
        }
    }
`;

class CreateItem extends Component {
    state = {
        title: '',
        description: '',
        image: '',
        largeImage: '',
        price: 0
    }

    handleChange = (e) => {
        const { name, type, value } = e.target;
        const val = type === 'number' ? parseFloat(value) : value;
        this.setState({ [name]: val });
    }

    uploadFile = async (e) => {
        const files = e.target.files;
        const data = new FormData();
        data.append('file', files[0]);
        data.append('upload_preset', 'sickfits');

        const res = await fetch('https://api.cloudinary.com/v1_1/dxfk2fjmp/image/upload', {
            method: 'POST',
            body: data
        });

        const file = await res.json();
        this.setState({
            image: file.secure_url,
            largeImage: file.eager[0].secure_url
        });
    }

    render() {
        return (
            <Mutation mutation={CREATE_ITEM_MUTATION} variables={this.state}>
                {(createItem, { loading, error }) => (
                    <Form data-test="CreateItem" onSubmit={async (e) => {
                        e.preventDefault();
                        // Send state to mutation
                        const res = await createItem();
                        const { data: {
                            createItem: { id } } } = res;
                        // Go to the new item page
                        Router.push({
                            pathname: '/item',
                            query: { id }
                        });
                    }}>
                        <ErrorMessage error={error} />
                        <fieldset disabled={loading} aria-busy={loading}>
                            <label htmlFor="file">
                                Image
                                <input
                                    type="file"
                                    id="file"
                                    name="file"
                                    placeholder="Upload an image"
                                    required
                                    onChange={this.uploadFile}
                                />
                                {this.state.image && <img src={this.state.image} alt={this.state.title} />}
                            </label>
                            <label htmlFor="title">
                                Title
                                <input
                                    type="text"
                                    id="title"
                                    name="title"
                                    placeholder="Title"
                                    required
                                    value={this.state.title}
                                    onChange={this.handleChange}
                                />
                            </label>
                            <label htmlFor="description">
                                Description
                                <textarea
                                    id="description"
                                    name="description"
                                    placeholder="Description"
                                    required
                                    value={this.state.description}
                                    onChange={this.handleChange}
                                />
                            </label>
                            <label htmlFor="price">
                                Price
                                <input
                                    type="number"
                                    id="price"
                                    name="price"
                                    placeholder="Price (cents)"
                                    required
                                    value={this.state.price}
                                    onChange={this.handleChange}
                                />
                            </label>
                            <button type="submit">Submit</button>
                        </fieldset>
                    </Form>
                )}
            </Mutation>
        );
    }

}

export default CreateItem;
