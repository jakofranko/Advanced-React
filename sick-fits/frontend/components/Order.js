import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Query } from 'react-apollo';
import { format } from 'date-fns';
import Head from 'next/head';
import gql from 'graphql-tag';
import formatMoney from '../lib/formatMoney';
import ErrorMessage from './ErrorMessage';
import OrderStyles from './styles/OrderStyles';

export const SINGLE_ORDER_QUERY = gql`
    query SINGLE_ORDER_QUERY($id: ID!) {
        order(id: $id) {
            id
            charge
            total
            createdAt
            user {
                id
            }
            items {
                id
                title
                description
                price
                image
                quantity
            }
        }
    }
`;

class Order extends Component {
    static propTypes = {
        id: PropTypes.string.isRequired
    }
    render() {
        return (
            <Query
                query={SINGLE_ORDER_QUERY}
                variables={{
                    id: this.props.id
                }}
            >
                {({ data, error, loading }) => {
                    if (error) return <ErrorMessage error={error} />;
                    if (loading) return <p>Loading...</p>;

                    const { order } = data;
                    return (
                        <OrderStyles data-test="Order">
                            <Head>
                                <title>Sick Fits - Order {this.props.id}</title>
                            </Head>
                            <p>
                                <span>Order ID:</span>
                                <span>{this.props.id}</span>
                            </p>
                            <p>
                                <span>Charge</span>
                                <span>{order.charge}</span>
                            </p>
                            <p>
                                <span>Date</span>
                                <span>{format(order.createdAt, 'MMMM d, YYYY h:mm a')}</span>
                            </p>
                            <p>
                                <span>Total</span>
                                <span>{formatMoney(order.total)}</span>
                            </p>
                            <p>
                                <span>Item Count</span>
                                <span>{order.items.length}</span>
                            </p>
                            <div className="items">
                                {order.items.map(orderItem => {
                                    return (
                                        <div key={orderItem.id} className="order-item">
                                            <img src={orderItem.image} alt={orderItem.title} />
                                            <div className="item-details">
                                                <h2>{orderItem.title}</h2>
                                                <p>Qty: {orderItem.quantity}</p>
                                                <p>Each: {formatMoney(orderItem.price)}</p>
                                                <p>SubTotal: {formatMoney(orderItem.price * orderItem.quantity)}</p>
                                            </div>
                                        </div>
                                    )
                                })}
                            </div>
                        </OrderStyles>
                    );
                }}
            </Query>
        );
    }

}

export default Order;
