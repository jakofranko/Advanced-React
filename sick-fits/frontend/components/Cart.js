import React, { Component } from 'react';
import { Query, Mutation } from 'react-apollo';
import { adopt } from 'react-adopt';
import gql from 'graphql-tag';
import CartStyles from './styles/CartStyles';
import Supreme from './styles/Supreme';
import CloseButton from './styles/CloseButton';
import SickButton from './styles/SickButton';
import User from './User';
import CartItem from './CartItem';
import TakeMyMoney from './TakeMyMoney';
import calcTotalPrice from '../lib/calcTotalPrice';
import formatMoney from '../lib/formatMoney';

export const LOCAL_STATE_QUERY = gql`
    query LOCAL_STATE_QUERY {
        cartOpen @client
    }
`;

export const TOGGLE_CART_MUTATION = gql`
    mutation TOGGLE_CART_MUTATION {
        toggleCart @client
    }
`;

const Composed = adopt({
    user: ({ render }) => <User>{render}</User>,
    toggleCart: ({ render }) => <Mutation mutation={TOGGLE_CART_MUTATION}>{render}</Mutation>,
    localState: ({ render }) => <Query query={LOCAL_STATE_QUERY}>{render}</Query>
});

class Cart extends Component {
    render() {
        return (
            <Composed>
                {({ user, toggleCart, localState}) => {
                    const { me } = user.data;
                    const { cartOpen } = localState.data;
                    if (!me) return null;
                    return (
                        <CartStyles open={cartOpen}>
                            <header>
                                <CloseButton onClick={toggleCart}>&times;</CloseButton>
                                <Supreme>{me.name}'s Cart</Supreme>
                                <p>You have {me.cart.length} item{me.cart.length === 1 ? '' : 's'} in your cart.</p>
                            </header>
                            <ul>
                                {me.cart.map(cartItem => (
                                    <CartItem cartItem={cartItem} key={cartItem.id} />
                                ))}
                            </ul>
                            <footer>
                                <p>{formatMoney(calcTotalPrice(me.cart))}</p>
                                {me.cart.length && <TakeMyMoney>
                                    <SickButton>Checkout</SickButton>
                                </TakeMyMoney>}
                            </footer>
                        </CartStyles>
                    );
                }}
            </Composed>
        );
    }

}

export default Cart;
