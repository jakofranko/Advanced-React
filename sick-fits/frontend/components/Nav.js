import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import Link from 'next/link';
import NavStyles from './styles/NavStyles';
import User, { CURRENT_USER_QUERY } from './User';
import { TOGGLE_CART_MUTATION } from './Cart';
import CartCount from './CartCount';

export const SIGNOUT_MUTATION = gql`
    mutation SIGNOUT_MUTATION {
        signout {
            message
        }
    }
`;

export default function Nav() {
    return (
        <User>
            {({ data: { me } }) => (
                <NavStyles>
                    <Link href="/items">
                        <a>Shop</a>
                    </Link>
                    {me && (
                        <>
                            <Link href="/sell">
                                <a>Sell</a>
                            </Link>
                            <Link href="/orders">
                                <a>Orders</a>
                            </Link>
                            <Link href="/me">
                                <a>Account</a>
                            </Link>
                            <Mutation mutation={TOGGLE_CART_MUTATION}>
                                {(toggleCart) =>(
                                    <button onClick={toggleCart}>
                                        My Cart
                                        <CartCount count={me.cart.reduce((tally, cartItem) => tally + cartItem.quantity, 0)} />
                                    </button>
                                )}
                            </Mutation>
                            <Mutation
                                mutation={SIGNOUT_MUTATION}
                                refetchQueries={[{ query: CURRENT_USER_QUERY }]}
                            >
                                {(signout) => (
                                    <button
                                        onClick={async (e) => {
                                            e.preventDefault();
                                            await signout();
                                        }}
                                    >
                                        Sign Out
                                    </button>
                                )}
                            </Mutation>
                        </>
                    )}
                    {!me && (
                        <Link href="/signup">
                            <a>Sign In</a>
                        </Link>
                    )}
                </NavStyles>
            )}
        </User>
    );
}
