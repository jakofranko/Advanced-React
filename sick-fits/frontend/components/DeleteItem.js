import React, { Component } from 'react';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import { ALL_ITEMS_QUERY } from './Items';

export const DELETE_ITEM_MUTATION = gql`
    mutation DELETE_ITEM_MUTATION($id: ID!) {
        deleteItem(id: $id) {
            id
        }
    }
`;

class DeleteItem extends Component {
    update = (cache, payload) => {
        // manually update cache so it matches server
        // 1. find the items we want in the cache
        const data = cache.readQuery({ query: ALL_ITEMS_QUERY });
        // 2. filter the deleted item out of the cache
        data.items = data.items.filter(i => i.id !== payload.data.deleteItem.id);
        // 3. put items back
        cache.writeQuery({ query: ALL_ITEMS_QUERY, data });

    }
    render() {
        return (
            <Mutation
                mutation={DELETE_ITEM_MUTATION}
                variables={{ id: this.props.id }}
                update={this.update}
            >
                {(deleteItem, { error }) => (
                    <button onClick={() => {
                        if (confirm('Are you sure you want to delete this item?')) {
                            deleteItem()
                                .catch(err => alert(err));
                        }
                    }}>
                        Delete item
                    </button>
                )}
            </Mutation>
        );
    }

}

export default DeleteItem;
