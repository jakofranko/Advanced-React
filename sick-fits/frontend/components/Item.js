import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Link from 'next/link';
import Title from './styles/Title';
import ItemStyles from './styles/ItemStyles';
import PriceTag from './styles/PriceTag';
import AddToCart from './AddToCart';
import DeleteItem from './DeleteItem';
import formatMoney from '../lib/formatMoney';

const canUpdatePerms = ['ADMIN', 'ITEMUPDATE'];
const canDeletePerms = ['ADMIN', 'ITEMDELETE'];

export default class Item extends Component {
    static propTypes = {
        item: PropTypes.shape({
            title: PropTypes.string.isRequired,
            description: PropTypes.string.isReqruied,
        }),
        userPermissions: PropTypes.arrayOf(PropTypes.string)
    }
    static defaultProps = {
        userPermissions: []
    }
    render() {
        const { item, userPermissions } = this.props;
        const canUpdate = canUpdatePerms.some(p => userPermissions.includes(p));
        const canDelete = canDeletePerms.some(p => userPermissions.includes(p));

        return (
            <ItemStyles>
                {item.image && <img src={item.image} alt={item.title}/>}

                <Title>
                    <Link href={{
                        pathname: '/item',
                        query: { id: item.id }
                    }}>
                        <a>{item.title}</a>
                    </Link>
                </Title>
                <PriceTag>{formatMoney(item.price)}</PriceTag>
                <p>{item.description}</p>
                <div className="buttonList">
                    {canUpdate && (
                        <Link href={{
                            pathname: "/update",
                            query: { id: item.id }
                        }}>
                            <a>Edit ‚úè</a>
                        </Link>
                    )}
                    <AddToCart id={item.id} />
                    {canDelete && (
                        <DeleteItem id={item.id} />
                    )}
                </div>
            </ItemStyles>
        )
    }
}
