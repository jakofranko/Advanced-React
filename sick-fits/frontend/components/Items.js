import React, { Component } from 'react';
import { Query } from 'react-apollo';
import { adopt } from 'react-adopt';
import gql from 'graphql-tag';
import styled from 'styled-components';
import Item from '../components/Item';
import User from './User';
import ErrorMessage from './ErrorMessage';
import Pagination from '../components/Pagination';
import { perPage } from '../config';

export const ALL_ITEMS_QUERY = gql`
    query ALL_ITEMS_QUERY($skip: Int = 0, $first: Int = ${perPage}) {
        items(first: $first, skip: $skip, orderBy: createdAt_DESC) {
            id
            title
            price
            description
            image
            largeImage
        }
    }
`;

const Composed = adopt({
    user: ({ render }) => <User>{render}</User>,
    items: ({ page, render }) => (
        <Query
            query={ALL_ITEMS_QUERY}
            variables={{ skip: page * perPage - perPage }}
        >
            {render}
        </Query>
    )
});

const Center = styled.div`
    text-align: center;
`;

const ItemsList = styled.div`
    display: grid;
    grid-template-columns: 1fr 1fr;
    grid-gap: 60px;
    max-width: ${props => props.theme.maxWidth};
    margin: 0 auto;
`;

export default class Items extends Component {
    render() {
        return (
            <Center>
                <Composed page={this.props.page} >
                    {({ user, items }) => {
                        const { data: itemsData } = items;
                        const { data: { me = {} } } = user;
                        const { permissions } = me;

                        if (user.loading || items.loading) return <p>Loading...</p>;
                        if (user.error) return <ErrorMessage error={user.error} />;
                        if (items.error) return <ErrorMessage error={items.error} />;

                        return (
                            <div>
                                <Pagination page={this.props.page} />
                                <p>I found {itemsData.items.length} items</p>
                                <ItemsList>
                                    {itemsData.items.map(item => (
                                        <Item
                                            userPermissions={permissions}
                                            item={item}
                                            key={item.id}
                                        />
                                    ))}
                                </ItemsList>
                                <Pagination page={this.props.page} />
                            </div>
                        );
                    }}
                </Composed>
            </Center>
        );
    }
}
