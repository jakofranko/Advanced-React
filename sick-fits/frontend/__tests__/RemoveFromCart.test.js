import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import wait from 'waait';
import { ApolloConsumer } from 'react-apollo';
import RemoveFromCart, { REMOVE_FROM_CART_MUTATION } from '../components/RemoveCartItem';
import { CURRENT_USER_QUERY } from '../components/User';
import { MockedProvider } from 'react-apollo/test-utils';
import { fakeUser, fakeCartItem } from '../lib/testUtils';

global.alert = console.log;

const cartItemId = 'unittestcartitemid';
const mocks = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: {
      data: {
        me: {
          ...fakeUser(),
          cart: [fakeCartItem({ id: cartItemId, price: 5000 })]
        }
      }
    }
  },
  {
    request: {
      query: REMOVE_FROM_CART_MUTATION,
      variables: { id: cartItemId }
    },
    result: {
      data: {
        removeFromCart: {
          __typename: 'CartItem',
          id: cartItemId
        }
      }
    }
  }
];

describe('<RemoveFromCart />', () => {
  it('should render and match snapshot', () => {
    const comp = mount(
      <MockedProvider>
        <RemoveFromCart id={cartItemId} />
      </MockedProvider>
    );

    expect(toJSON(comp.find('button'))).toMatchSnapshot();
  });

  it('should remove the item from the cart', async () => {
    let apolloClient;
    const comp = mount(
      <MockedProvider mocks={mocks}>
        <ApolloConsumer>
          {client => {
            apolloClient = client;
            return <RemoveFromCart id={cartItemId} />
          }}
        </ApolloConsumer>
      </MockedProvider>
    );

    const { data: { me } } = await apolloClient.query({ query: CURRENT_USER_QUERY });
    expect(me.cart).toHaveLength(1);
    expect(me.cart[0].item.price).toBe(5000);

    // remove item
    comp.find('button').simulate('click');
    await wait();

    const { data: { me: me2 } } = await apolloClient.query({ query: CURRENT_USER_QUERY });
    expect(me2.cart).toHaveLength(0);    
  });
});
