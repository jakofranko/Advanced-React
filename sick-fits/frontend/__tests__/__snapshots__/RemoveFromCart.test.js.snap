// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RemoveFromCart /> should render and match snapshot 1`] = `
<button
  className="RemoveCartItem__BigButton-sc-1mkopzv-0 dpUZfW"
  disabled={false}
  onClick={[Function]}
  title="Delete Item"
>
  Ã—
</button>
`;
